pkg/controller/datavolume-controller.go

```diff
@@ -595,24 +553,194 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
+		}
+
+		if pvc.Status.Phase == corev1.ClaimBound {
+			if v, ok := pvc.Annotations[AnnCloneOf]; !ok || v != "true" {
+				if pvc.Annotations == nil {
+					pvc.Annotations = make(map[string]string)
+				}
+				pvc.Annotations[AnnCloneOf] = "true"
+
+				if err := r.client.Update(context.TODO(), pvc); err != nil {
+					return reconcile.Result{}, err
+				}
 			}
+		}
+	}
+
+	return reconcile.Result{Requeue: true}, r.updateCloneStatusPhase(cdiv1.CSICloneInProgress, datavolume, nil, CsiClone)
+}
+
+// When the clone is finished some additional actions may be applied
+// like namespaceTransfer Cleanup or size expansion
+func (r *DatavolumeReconciler) finishClone(log logr.Logger,
+	datavolume *cdiv1.DataVolume,
+	pvc *corev1.PersistentVolumeClaim,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	transferName string,
+	selectedCloneStrategy cloneStrategy) (reconcile.Result, error) {
+
+	if isCrossNamespaceClone(datavolume) && datavolume.Status.Phase == cdiv1.Succeeded {
+		if err := r.cleanupTransfer(log, datavolume, transferName); err != nil {
+			return reconcile.Result{}, err
+		}
 
+		// done, done
+		return reconcile.Result{}, nil
+	}
+
+	//DO Nothing, not yet ready
+	if pvc.Annotations[AnnCloneOf] != "true" {
+		// TODO: BR: I think a Status update here might help with some edge cases, investigation in progress
+		//  r.updateCloneStatusPhase(cdiv1.CSICloneInProgress, datavolume, nil, "csivolumeclone")
+		return reconcile.Result{}, nil
+	}
+
+	// expand for non-namespace case
+	return r.expandPvcAfterClone(log, datavolume, pvc, pvcSpec, selectedCloneStrategy)
+}
+
+func (r *DatavolumeReconciler) expandPvcAfterClone(log logr.Logger,
+	datavolume *cdiv1.DataVolume,
+	pvc *corev1.PersistentVolumeClaim,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	selectedCloneStrategy cloneStrategy) (reconcile.Result, error) {
+	done, err := r.expand(log, datavolume, pvc, pvcSpec)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if !done {
+		return reconcile.Result{}, r.updateCloneStatusPhase(cdiv1.ExpansionInProgress, datavolume, pvc, selectedCloneStrategy)
+	}
+
+	// done
+	return reconcile.Result{}, r.updateCloneStatusPhase(cdiv1.Succeeded, datavolume, pvc, selectedCloneStrategy)
+}
+
+// temporary pvc is used when the clone src and tgt are in two distinct namespaces
+func (r *DatavolumeReconciler) expandTmpPvcAfterClone(log logr.Logger,
+	datavolume *cdiv1.DataVolume,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	tmpPVC *corev1.PersistentVolumeClaim,
+	selectedCloneStrategy cloneStrategy) (reconcile.Result, error) {
+	done, err := r.expand(log, datavolume, tmpPVC, pvcSpec)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !done {
+		return reconcile.Result{Requeue: true},
+			r.updateCloneStatusPhase(cdiv1.ExpansionInProgress, datavolume, nil, selectedCloneStrategy)
+	}
+
+	// trigger transfer and next reconcile should have pvcExists == true
+	tmpPVC.Annotations[annReadyForTransfer] = "true"
```

Comment:
expandTmpPvcAfterClone and expandPvcAfterClone look very similar, except for some additional annotations and the updateCloneStatusPhase is different. I think maybe we can combine some of that?

Reply:
I did try extracting the expand:
```
func (r *DatavolumeReconciler) expandPvcAfterClone(log logr.Logger,
	datavolume *cdiv1.DataVolume,
	pvc *corev1.PersistentVolumeClaim,
	pvcSpec *corev1.PersistentVolumeClaimSpec,
	selectedCloneStrategy cloneStrategy) (bool, error) {
	done, err := r.expand(log, datavolume, pvc, pvcSpec)
	if err != nil {
		return false, err
	}
	if !done {
		return false, r.updateCloneStatusPhase(cdiv1.ExpansionInProgress, datavolume, pvc, selectedCloneStrategy)
	}

	// done
	return true, nil
}
```

But it does not look much better in my opinion.  Any ideas?